modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1317,152 +1370,77 @@ private RuntimeException convertThrowable(Throwable th) {
-                    changePeersQueue.clear();
+                List<ClusterNode> newPeers = ((List<ClusterNode>) ByteUtils.fromBytes(evt.entryEvent().newEntry().value()));
 
-                    for (int i = 0; i < tbls.size(); i++) {
-                        tbls.createOrUpdate(tbls.get(i).name(), changeX -> {
-                            ExtendedTableChange change = (ExtendedTableChange) changeX;
-                            byte[] currAssignments = change.assignments();
+                RaftGroupService raftGrpSvc = null;
 
-                            List<List<ClusterNode>> recalculatedAssignments = AffinityUtils.calculateAssignments(
-                                    clusterNodes,
-                                    change.partitions(),
-                                    change.replicas());
+                try {
+                    var deltaPeers = newPeers.stream()
+                            .filter(p -> !assignments.get(part).contains(p))
+                            .collect(Collectors.toList());
 
-                            if (!recalculatedAssignments.equals(ByteUtils.fromBytes(currAssignments))) {
-                                change.changeAssignments(ByteUtils.toBytes(recalculatedAssignments));
+                    raftGrpSvc = raftMgr.updateRaftGroup(grpId, assignments.get(part), deltaPeers, raftGrpLsnrSupplier,
+                            raftGrpEvtsLsnrSupplier).join();
 
-                                changePeersQueue.add(() ->
-                                        updateRaftTopology(
-                                                (List<List<ClusterNode>>) ByteUtils.fromBytes(currAssignments),
-                                                recalculatedAssignments,
-                                                change.id()));
-                            }
-                        });
-                    }
-                })
-                .thenCompose((v) -> {
-                    CompletableFuture<?>[] changePeersFutures = new CompletableFuture<?>[changePeersQueue.size()];
+                    raftGrpSvc.refreshLeader().join();
 
-                    int i = 0;
+                    if (new Peer(raftGrpSvc.clusterService().topologyService().localMember().address()).equals(raftGrpSvc.leader())) {
 
-                    for (Supplier<CompletableFuture<Void>> task : changePeersQueue) {
-                        changePeersFutures[i++] = task.get();
-                    }
+                        var newNodes = newPeers.stream().map(n -> new Peer(n.address())).collect(Collectors.toList());
 
-                    return CompletableFuture.allOf(changePeersFutures);
-                })
-                .whenComplete((res, th) -> {
-                    if (th != null) {
-                        setBaselineFut.completeExceptionally(th);
-                    } else {
-                        setBaselineFut.complete(null);
+                        raftGrpSvc.changePeersAsync(newNodes).join();
```

Comment:
Well, the core idea was - not to start changePeers process for stale notifications. Seems that it's not true right now, am I wrong?

Reply:
updated according to our discussion